lib/translate/status_translator.h

```diff
@@ -0,0 +1,46 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIB_STATUS_TRANSLATOR_H
+#define LIB_STATUS_TRANSLATOR_H
+
+#include <optional>
+
+#include "../../third_party/spdk_defs/nvme_defs.h"
+#include "../scsi_defs.h"
+
+namespace translator {
+
+struct TranslatedScsiStatus {
+  scsi_defs::Status status;
+  scsi_defs::SenseKey sense_key;
+  std::optional<scsi_defs::AdditionalSenseCode> sense_code;
+
+  TranslatedScsiStatus() {}
+  TranslatedScsiStatus(scsi_defs::Status status, scsi_defs::SenseKey sense_key,
+                       std::optional<scsi_defs::AdditionalSenseCode> sense_code)
+      : status{status}, sense_key{sense_key}, sense_code{sense_code} {}
+};
+
+int translateStatusCode(const uint8_t sct, const uint8_t sc,
+                        TranslatedScsiStatus *translated);
+int translateStatusCode(const nvme_defs::GenericCommandStatusCode sc,
+                        TranslatedScsiStatus *translated);
+int translateStatusCode(const nvme_defs::CommandSpecificStatusCode sc,
+                        TranslatedScsiStatus *translated);
+int translateStatusCode(const nvme_defs::MediaErrorStatusCode sc,
+                        TranslatedScsiStatus *translated);
```

Comment:
@Ralith is `std::optional` easily available in kernel modules?

Reply:
That's a good question; if it's not, the definition of `TranslatedScsiStatus` will need to be adjusted as well. I suspect we might get away with it since, being a template, it's unlikely to pull in any standard library runtime symbols, but some research and/or experimentation will be needed to confirm.

Reply:
#71 for follow-ups on this discussion